### PR TITLE
Move many nuclear data-related classes to C++

### LINKED ISSUE #1017: Decide on library for N-dimensional arrays
With plenty of action on #603 these days, we will be needing a clean way to express N-dimensional arrays in C++ which the language itself doesn't really provide. There are a number of libraries that provide multi-dimensional arrays, including:
- [Boost](https://www.boost.org/doc/libs/1_62_0/libs/multi_array/doc/user.html)
- [Eigen](http://eigen.tuxfamily.org/index.php?title=Main_Page) (intended mostly for 2D only, although they have an experimental N-d array)
- [Armadillo](http://arma.sourceforge.net/) (again, primarily for matrices)
- [DyND](http://libdynd.org/) (doesn't seem very mature or actively developed?)
- [xtensor](https://xtensor.readthedocs.io/en/latest/)
- [kokkos array_ref](https://github.com/kokkos/array_ref)

Looking a little bit at all the options, I'm currently a big fan of xtensor. It has a beautiful, numpy-inspired API that will ease the learning curve for new developers. It is header-only so we should be able to just directly include it in our source tree the same way we handle pugixml. It does require some C++14 features and would require we bump our gcc dependency to 4.9, but I think I'd be willing to live with that. I'm curious to hear others thoughts.

### PR DESCRIPTION
**Pull request by paulromano**
### Overview

This pull request moves most of the nuclear data class hierarchy to C++. Classes that have been moved over include:

- One-dimensional functions (`Tabulated1D`, `Polynomial`)
- Univariate probability distributions (`Discrete`, `Uniform`, `Maxwell`, etc.)
- Pure angular distributions (`PolarAzimuthal`, `Isotropic`, etc.)
- Spatial distributions (`CartesianIndependent`, `SpatialBox`, etc.)
- Angle distributions dependent on incident energy (`AngleDistribution`)
- Energy distributions dependent on incident energy (`LevelInelastic`, `ContinuousTabular`, etc.)
- Correlated and uncorrelated angle-energy distributions (`UncorrelatedAngleEnergy`, `KalbachMann`, etc.)
- Reaction products (`ReactionProduct`)
- Reactions (`Reaction`)

The connection between C++ and Fortran happens at the level of the `Reaction` type, so you'll see some glue code there for accessing reaction cross sections, product yields, sampling product distributions, and so on. Everything below the `Reaction` happens on the C++ side now.

While I was able to make a net reduction of a bit more than 1k lines of Fortran code, some of the classes couldn't be completely removed because they are used elsewhere in the Fortran code. Namely,

- `CorrelatedAngleEnergy` is used when reading thermal scattering data from HDF5 files (even though the type itself is not ultimately used to store the data)
- Univariate and multivariate probability distributions are also used for external source distributions, so those are still around on the Fortran side
- `Tabulated1D` is used in a bunch of other places, so that couldn't be removed yet

Once those can removed, that's about another 1k LOC.

### xtensor

This PR relies on the [xtensor](https://xtensor.readthedocs.io/en/latest/) library which provides a clean API for dealing with multidimensional arrays. As with pugixml, I've included the headers directly in our source tree. After a bit of a learning curve, I'm really enjoying using it. The one potential downside, discussed in #1017, is that it requires C++14 (gcc 4.9). While I'm in favor of allowing C++14 code, the fact that it requires gcc 4.9 causes some problems with our CI testing because Travis CI is still based on Ubuntu 14.04, which only has gcc 4.8. The options to get around this are to:

1. Install a newer gcc from the `ubuntu-toolchain-r-test` PPA. This is really straightforward (in fact, this is what the NJOY team does) but for us the problem is that it would force us back into the mode of installing HDF5/MPICH from source rather than relying on the default repositories.
2. Use docker. I've never used it, so going down that path will require a bit of investment.
3. Evidently Travis already has a build environment for Ubuntu 16.04 that is [experimental/unsupported](https://github.com/travis-ci/travis-ci/issues/7260#issuecomment-354350596) that can be enabled by just specifying `dist: xenial` in .travis.yml. It looks like the version of gcc installed on this environment is 5.4, so plenty sufficient for our needs.

I opted to go for option 3 since it seems like the path of least resistence, and the Travis team has been saying for a while now that support for 16.04 is "on the roadmap" and "coming soon". One issue I ran into using `dist: xenial` is that it doesn't work with Python 3.4. Given that the end-of-life for Python 3.4 is [seven months away](https://devguide.python.org/#status-of-python-branches), I'm not too concerned if we're not testing it at this point. Plus, moving to xenial actually means we can start testing Python 3.7 (which *only* works there).

Looks like a full test configuration on the Travis xenial build environment is about 4 min longer on average than trusty, so slightly slower but not prohibitive. At some point, we may want to look at the container-based infrastructure on Travis as it might provide better build times.

Closes #1017.

### NumPy test failures

When I enabled Python 3.7 in the Travis build, it had some problems installing numpy because we were requiring a version less than 1.14 for which there are no wheels on PyPI for Python 3.7. I've decided to just update our test results to work with numpy 1.14 and up (and not with version 1.13 and before) and added a version check when running regression tests to make sure a suitable version is being used.

### CODE REVIEW

File: src/xml_interface.cpp [hunk 1]
```diff
@@ -23,9 +23,15 @@ get_node_value(pugi::xml_node node, const char *name)
...
     fatal_error(err_msg);
   }
+  return value_char;
+}
+
 
-  // Convert to lowercase string.
-  std::string value(value_char);
+std::string
+get_node_value(pugi::xml_node node, const char *name)
```
Comment (smharper) [inline src/xml_interface.cpp]: `char *name` --> `char* name`

File: src/xml_interface.cpp [hunk 2]
```diff
@@ -9,7 +9,7 @@
 namespace openmc {
 
 std::string
-get_node_value(pugi::xml_node node, const char *name)
+get_node_value_str(pugi::xml_node node, const char* name)
```
Comment (smharper) [inline src/xml_interface.cpp]: Why did you need to make this function?  If it's just because of the lowercasing, I would rather there was an optional argument to stop the lowercasing (we need it for `<name>` tags anyway).

File: src/cell.cpp [hunk 3]
```diff
@@ -15,18 +14,6 @@
...
 
-//==============================================================================
-// Constants
-//==============================================================================
-
-constexpr int32_t OP_LEFT_PAREN   {std::numeric_limits<int32_t>::max()};
-constexpr int32_t OP_RIGHT_PAREN  {std::numeric_limits<int32_t>::max() - 1};
-constexpr int32_t OP_COMPLEMENT   {std::numeric_limits<int32_t>::max() - 2};
-constexpr int32_t OP_INTERSECTION {std::numeric_limits<int32_t>::max() - 3};
-constexpr int32_t OP_UNION        {std::numeric_limits<int32_t>::max() - 4};
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/cell.cpp]: True; I'll move them back.

File: src/xml_interface.cpp (hunk 2 above)
Comment (paulromano) [reply src/xml_interface.cpp]: Yeah, there were some cases where we were lowercasing and others where we were not. I'll combine these two functions into a single one with optional arguments and only use lowercasing where it is necessary.

**Comment by smharper** [comment]
Thanks @paulromano.  Merging!
